In [1]:
import matplotlib.pyplot as plt
import numpy as np
from sklearn.model_selection import learning_curve, validation_curve


def plot_learning_curve(
    estimator,
    X,
    y,
    title="Learning Curve",
    ylim=None,
    cv=None,
    n_jobs=None,
    train_sizes=np.linspace(0.1, 1.0, 5),
    scoring=None,
):
    """
    This is a custom modification of the code present here:
    https://scikit-learn.org/stable/auto_examples/model_selection/plot_learning_curve.html
    
    Generate 3 plots: the test and training learning curve, the training
    samples vs fit times curve, the fit times vs score curve.

    Parameters
    ----------
    estimator : object type that implements the "fit" and "predict" methods
        An object of that type which is cloned for each validation.

    title : string
        Title for the chart.

    X : array-like, shape (n_samples, n_features)
        Training vector, where n_samples is the number of samples and
        n_features is the number of features.

    y : array-like, shape (n_samples) or (n_samples, n_features), optional
        Target relative to X for classification or regression;
        None for unsupervised learning.

    ylim : tuple, shape (ymin, ymax), optional
        Defines minimum and maximum yvalues plotted.

    cv : int, cross-validation generator or an iterable, optional
        Determines the cross-validation splitting strategy.
        Possible inputs for cv are:
          - None, to use the default 5-fold cross-validation,
          - integer, to specify the number of folds.
          - :term:`CV splitter`,
          - An iterable yielding (train, test) splits as arrays of indices.

        For integer/None inputs, if ``y`` is binary or multiclass,
        :class:`StratifiedKFold` used. If the estimator is not a classifier
        or if ``y`` is neither binary nor multiclass, :class:`KFold` is used.

        Refer :ref:`User Guide <cross_validation>` for the various
        cross-validators that can be used here.

    n_jobs : int or None, optional (default=None)
        Number of jobs to run in parallel.
        ``None`` means 1 unless in a :obj:`joblib.parallel_backend` context.
        ``-1`` means using all processors. See :term:`Glossary <n_jobs>`
        for more details.

    train_sizes : array-like, shape (n_ticks,), dtype float or int
        Relative or absolute numbers of training examples that will be used to
        generate the learning curve. If the dtype is float, it is regarded as a
        fraction of the maximum size of the training set (that is determined
        by the selected validation method), i.e. it has to be within (0, 1].
        Otherwise it is interpreted as absolute sizes of the training sets.
        Note that for classification the number of samples usually have to
        be big enough to contain at least one sample from each class.
        (default: np.linspace(0.1, 1.0, 5))

    scoring: string, callable or None, optional, default: None
        A string (see model evaluation documentation) or a scorer callable object / function
        with signature scorer(estimator, X, y).
    """

    fig, axes = plt.subplots(1, 1, figsize=(10, 5))

    axes.set_title(title)
    if ylim is not None:
        axes.set_ylim(*ylim)
    axes.set_xlabel("Training examples")
    axes.set_ylabel("Score")

    train_sizes, train_scores, test_scores = learning_curve(
        estimator, X, y, cv=cv, n_jobs=n_jobs, train_sizes=train_sizes, scoring=scoring
    )
    train_scores_mean = np.mean(train_scores, axis=1)
    train_scores_std = np.std(train_scores, axis=1)
    test_scores_mean = np.mean(test_scores, axis=1)
    test_scores_std = np.std(test_scores, axis=1)

    # Plot learning curve
    axes.grid(True)
    axes.fill_between(
        train_sizes,
        train_scores_mean - train_scores_std,
        train_scores_mean + train_scores_std,
        alpha=0.1,
        color="r",
    )
    axes.fill_between(
        train_sizes,
        test_scores_mean - test_scores_std,
        test_scores_mean + test_scores_std,
        alpha=0.1,
        color="g",
    )
    axes.plot(train_sizes, train_scores_mean, "o-", color="r", label="Training score")
    axes.plot(
        train_sizes, test_scores_mean, "o-", color="g", label="Cross-validation score"
    )
    axes.legend(loc="best")
    return fig, axes


def plot_validation_curve(
    estimator,
    X,
    y,
    ylim=None,
    cv=None,
    n_jobs=None,
    param_name=None,
    param_range=None,
    scoring=None,
):
    """
    referred from :
    https://scikit-learn.org/stable/modules/model_evaluation.html#classification-metrics

    :param estimator: object type that implements the "fit" and "predict" methods
        An object of that type which is cloned for each validation.
    :param X: array-like, shape (n_samples, n_features)
        Training vector, where n_samples is the number of samples and
        n_features is the number of features.
    :param y: array-like, shape (n_samples) or (n_samples, n_features), optional
        Target relative to X for classification or regression;
        None for unsupervised learning.
    :param ylim: tuple, shape (ymin, ymax), optional
        Defines minimum and maximum yvalues plotted.
    :param cv: int, cross-validation generator or an iterable, optional
        Determines the cross-validation splitting strategy.
        Possible inputs for cv are:
          - None, to use the default 5-fold cross-validation,
          - integer, to specify the number of folds.
          - :term:`CV splitter`,
          - An iterable yielding (train, test) splits as arrays of indices.

        For integer/None inputs, if ``y`` is binary or multiclass,
        :class:`StratifiedKFold` used. If the estimator is not a classifier
        or if ``y`` is neither binary nor multiclass, :class:`KFold` is used.

        Refer :ref:`User Guide <cross_validation>` for the various
        cross-validators that can be used here.
    :param n_jobs: int or None, optional (default=None)
        Number of jobs to run in parallel.
        ``None`` means 1 unless in a :obj:`joblib.parallel_backend` context.
        ``-1`` means using all processors. See :term:`Glossary <n_jobs>`
        for more details.
    :param param_range: array-like, shape (n_values,)
    The values of the parameter that will be evaluated.
    :param param_name: string
    Name of the parameter that will be varied.
    :param scoring: string, callable or None, optional, default: None
        A string (see model evaluation documentation) or a scorer callable object / function
        with signature scorer(estimator, X, y).
    :return: fig


    """
    train_scores, test_scores = validation_curve(
        estimator,
        X,
        y,
        param_name=param_name,
        param_range=param_range,
        scoring=scoring,
        n_jobs=n_jobs,
        cv=cv,
    )
    train_scores_mean = np.mean(train_scores, axis=1)
    train_scores_std = np.std(train_scores, axis=1)
    test_scores_mean = np.mean(test_scores, axis=1)
    test_scores_std = np.std(test_scores, axis=1)

    fig, axes = plt.subplots(1, 1, figsize=(10, 5))
    axes.grid(True)
    axes.set_title(f"Validation Curve with {estimator.__class__}")
    axes.set_xlabel(f"{param_name}")
    axes.set_ylabel(f"{scoring}")
    if ylim is not None:
        axes.set_ylim(*ylim)

    lw = 2
    axes.plot(
        param_range,
        train_scores_mean,
        label="Training score",
        color="darkorange",
        lw=lw,
    )
    axes.fill_between(
        param_range,
        train_scores_mean - train_scores_std,
        train_scores_mean + train_scores_std,
        alpha=0.2,
        color="darkorange",
        lw=lw,
    )
    axes.plot(
        param_range,
        test_scores_mean,
        label="Cross-validation score",
        color="navy",
        lw=lw,
    )
    axes.fill_between(
        param_range,
        test_scores_mean - test_scores_std,
        test_scores_mean + test_scores_std,
        alpha=0.2,
        color="navy",
        lw=lw,
    )
    axes.legend(loc="best")
    return fig, axes


In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

from pathlib import Path

# classifiers we will use
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor, HistGradientBoostingRegressor
from sklearn.tree import DecisionTreeRegressor
import xgboost

#imputers
from sklearn.impute import SimpleImputer, KNNImputer

# model selection bits
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV, train_test_split, ParameterGrid, ParameterSampler
#from sklearn.model_selection import cross_val_score
from sklearn.model_selection import StratifiedShuffleSplit, GroupShuffleSplit, GroupKFold, StratifiedKFold
from sklearn.model_selection import learning_curve, validation_curve

# evaluation
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error


import scipy

Let's load all the data. 

In [3]:
# Load the competition datasets into Pandas DataFrame
path = Path("/Users/13392/Documents/amp-parkinsons-disease-progression-prediction")
proteins = pd.read_csv(path/"train_proteins.csv")
peptides = pd.read_csv(path/"train_peptides.csv")
clinical = pd.read_csv(path/"train_clinical_data.csv")
supplemental = pd.read_csv(path/"supplemental_clinical_data.csv")

As discussed previously, we are dropping the entire "medication status" column, because:
1) Over 50% values are NaN. 
2) the test dataset will not have this data. 

In [4]:
# drop the "medication status" column (due to over 50% NaN values), keep a copy of the original
# for later access. 
clinical_copy = clinical.copy()

clinical.drop('upd23b_clinical_state_on_medication', axis=1, inplace=True)

In [5]:

targets = ['updrs_1', 'updrs_2', 'updrs_3', 'updrs_4']
ids = ['patient_id', 'visit_id']
month = ['visit_month']

Let's see how much remains of 'NaN'

In [6]:
print(f'NaN value count:\n{clinical.isna().sum()}')
clinical

NaN value count:
visit_id          0
patient_id        0
visit_month       0
updrs_1           1
updrs_2           2
updrs_3          25
updrs_4        1038
dtype: int64


visit_id  patient_id  visit_month  updrs_1  updrs_2  updrs_3  updrs_4
0         55_0          55            0     10.0      6.0     15.0      NaN
1         55_3          55            3     10.0      7.0     25.0      NaN
2         55_6          55            6      8.0     10.0     34.0      NaN
3         55_9          55            9      8.0      9.0     30.0      0.0
4        55_12          55           12     10.0     10.0     41.0      0.0
...        ...         ...          ...      ...      ...      ...      ...
2610  65043_48       65043           48      7.0      6.0     13.0      0.0
2611  65043_54       65043           54      4.0      8.0     11.0      1.0
2612  65043_60       65043           60      6.0      6.0     16.0      1.0
2613  65043_72       65043           72      3.0      9.0     14.0      1.0
2614  65043_84       65043           84      7.0      9.0     20.0      3.0

[2615 rows x 7 columns]

Significant, but manageable with some kind of imputation. Let's count the number of visits each patient has on record. 

In [7]:
cols = ['patient_id', 'num_entries']
patient_list = clinical.patient_id.unique()
n_list = []
p_list = []

for patient in patient_list:
    n=len(clinical[clinical.patient_id==patient].index)

    n_list.append(n)
    p_list.append(patient)

df_visits_by_patient = pd.DataFrame(list(zip(p_list, n_list)), columns=cols)

df_visits_by_patient


patient_id  num_entries
0            55           13
1           942           15
2          1517           10
3          1923            7
4          2660            6
..          ...          ...
243       63875            9
244       63889           10
245       64669           15
246       64674           16
247       65043           15

[248 rows x 2 columns]

NaN values in the "Proteins"  and "Peptides" datasets. 

In [8]:
print(f'NaN value count:\n{proteins.isna().sum()}')
proteins

NaN value count:
visit_id       0
visit_month    0
patient_id     0
UniProt        0
NPX            0
dtype: int64


visit_id  visit_month  patient_id UniProt       NPX
0            55_0            0          55  O00391   11254.3
1            55_0            0          55  O00533  732430.0
2            55_0            0          55  O00584   39585.8
3            55_0            0          55  O14498   41526.9
4            55_0            0          55  O14773   31238.0
...           ...          ...         ...     ...       ...
232736  58648_108          108       58648  Q9UBX5   27387.8
232737  58648_108          108       58648  Q9UHG2  369437.0
232738  58648_108          108       58648  Q9UKV8  105830.0
232739  58648_108          108       58648  Q9Y646   21257.6
232740  58648_108          108       58648  Q9Y6R7   17953.1

[232741 rows x 5 columns]

In [9]:
print(f'NaN value count:\n{peptides.isna().sum()}')

peptides

NaN value count:
visit_id            0
visit_month         0
patient_id          0
UniProt             0
Peptide             0
PeptideAbundance    0
dtype: int64


visit_id  visit_month  patient_id UniProt  \
0            55_0            0          55  O00391   
1            55_0            0          55  O00533   
2            55_0            0          55  O00533   
3            55_0            0          55  O00533   
4            55_0            0          55  O00533   
...           ...          ...         ...     ...   
981829  58648_108          108       58648  Q9UHG2   
981830  58648_108          108       58648  Q9UKV8   
981831  58648_108          108       58648  Q9Y646   
981832  58648_108          108       58648  Q9Y6R7   
981833  58648_108          108       58648  Q9Y6R7   

                                       Peptide  PeptideAbundance  
0                                NEQEQPLGQWHLS          11254.30  
1                                  GNPEPTFSWTK         102060.00  
2                              IEIPSSVQQVPTIIK         174185.00  
3       KPQSAVYSTGSNGILLC(UniMod_4)EAEGEPQPTIK          27278.90  
4                                 SMEQNGPGLEYR          30838.70  
...                                        ...               ...  
981829                         ILAGSADSEGVAAPR         202820.00  
981830                           SGNIPAGTTVDTK         105830.00  
981831                             LALLVDTVGPR          21257.60  
981832        AGC(UniMod_4)VAESTAVC(UniMod_4)R           5127.26  
981833                          GATTSPGVYELSSR          12825.90  

[981834 rows x 6 columns]

Great! No NaN values at all in them. Let's find out how many of each patient's visits have protein/peptide data. 

In [10]:
cols = ['patient_id', 'num_entries_protein']
patient_list = proteins.patient_id.unique()
n_list = []
p_list = []

for patient in patient_list:
    n=len(proteins[proteins.patient_id==patient].visit_id.unique())

    n_list.append(n)
    p_list.append(patient)

df_recorded_visits_protein = pd.DataFrame(list(zip(p_list, n_list)), columns=cols)

df_recorded_visits_protein


patient_id  num_entries_protein
0            55                    4
1          1517                    4
2          1923                    3
3          2660                    5
4          3636                    3
..          ...                  ...
243       52998                    3
244       54979                    3
245       58597                    3
246        7508                    3
247       12931                    4

[248 rows x 2 columns]

In [11]:
cols = ['patient_id', 'num_entries_peptide']
patient_list = peptides.patient_id.unique()
n_list = []
p_list = []

for patient in patient_list:
    n=len(peptides[peptides.patient_id==patient].visit_id.unique())

    n_list.append(n)
    p_list.append(patient)

df_recorded_visits_peptide = pd.DataFrame(list(zip(p_list, n_list)), columns=cols)

df_recorded_visits_peptide

patient_id  num_entries_peptide
0            55                    4
1          1517                    4
2          1923                    3
3          2660                    5
4          3636                    3
..          ...                  ...
243       52998                    3
244       54979                    3
245       58597                    3
246        7508                    3
247       12931                    4

[248 rows x 2 columns]

In [12]:
df = pd.merge(df_recorded_visits_protein, df_recorded_visits_peptide, on='patient_id', how='left')
df = pd.merge(df_visits_by_patient, df, on='patient_id', how='left')
df.head(10)

patient_id  num_entries  num_entries_protein  num_entries_peptide
0          55           13                    4                    4
1         942           15                    4                    4
2        1517           10                    4                    4
3        1923            7                    3                    3
4        2660            6                    5                    5
5        3636           14                    3                    3
6        3863            9                    5                    5
7        4161           12                    6                    6
8        4172            8                    7                    7
9        4923           11                    5                    5

It's not looking great - we can clearly see that, for most of the patients, only 1/2 to 1/3 of the visits contain protein and peptide records.

Something to think about for later on...

Now we will "pivot" the datasets so the unique coding for each protein/peptide becomes a feature for the models to learn on. 

In [13]:
df_proteins = proteins.pivot(index=['patient_id', 'visit_month', 'visit_id'], columns='UniProt', values='NPX').rename_axis(columns=None).reset_index()

df_peptides = peptides.pivot(index=['patient_id', 'visit_month', 'visit_id'], columns='Peptide', values='PeptideAbundance').rename_axis(columns=None).reset_index()


In [14]:
df_proteins

patient_id  visit_month  visit_id   O00391     O00533   O00584   O14498  \
0             55            0      55_0  11254.3   732430.0  39585.8  41526.9   
1             55            6      55_6  13163.6   630465.0  35220.8  41295.0   
2             55           12     55_12  15257.6   815083.0  41650.9  39763.3   
3             55           36     55_36  13530.8   753832.0  43048.9  43503.6   
4            942            6     942_6  11218.7   399518.0  20581.0  31290.9   
...          ...          ...       ...      ...        ...      ...      ...   
1108       64674           84  64674_84      NaN   190487.0  24907.9  18543.1   
1109       65043            0   65043_0  13472.4   927954.0  42661.5  43663.2   
1110       65043           12  65043_12  14134.9   984651.0  28990.8  42440.9   
1111       65043           24  65043_24  14659.5  1062020.0  46440.4  38293.0   
1112       65043           48  65043_48  10589.6   902434.0  44890.8  38771.5   

        O14773   O14791    O15240  ...    Q9HDC9    Q9NQ79    Q9NYU2   Q9UBR2  \
0     31238.00  4202.71  177775.0  ...  365475.0  35528.00   97005.6  23122.5   
1     26219.90  4416.42  165638.0  ...  405676.0  30332.60  109174.0  23499.8   
2     30703.60  4343.60  151073.0  ...  303953.0  43026.20  114921.0  21860.1   
3     33577.60  5367.06  101056.0  ...  303597.0  48188.40  109794.0  23930.6   
4      6173.58  2564.37  160526.0  ...  253373.0  27431.80   93796.7  17450.9   
...        ...      ...       ...  ...       ...       ...       ...      ...   
1108  10124.90  2308.71   62095.4  ...  260021.0   7139.93  104277.0  10500.0   
1109  20071.30  3278.88  266339.0  ...  186414.0  25897.80       NaN  21480.7   
1110  25357.40  3267.66  270575.0  ...  301343.0  22343.40  105626.0  20500.8   
1111  21971.80  3990.34  221358.0  ...  300439.0  52143.60  139291.0  19449.2   
1112  20188.30  3137.07  203487.0  ...  223136.0  21659.90  116470.0  14073.3   

       Q9UBX5    Q9UHG2   Q9UKV8   Q9UNU6   Q9Y646   Q9Y6R7  
0     60912.6  408698.0      NaN  29758.8  23833.7  18953.5  
1     51655.8  369870.0      NaN  22935.2  17722.5  16642.7  
2     61598.2  318553.0  65762.6  29193.4  28536.1  19290.9  
3     70223.5  377550.0  74976.1  31732.6  22186.5  21717.1  
4     21299.1  306621.0  82335.5  24018.7  18939.5  15251.2  
...       ...       ...      ...      ...      ...      ...  
1108  21944.2  136725.0  62217.5      NaN  10287.7  13848.2  
1109  57364.0  416142.0  37584.6      NaN  28346.5  35617.5  
1110  54011.2  380072.0  40588.9      NaN  17035.7  37064.2  
1111  66569.9  300948.0  36150.4      NaN  21286.3  39587.9  
1112  48796.4  320821.0  39046.7      NaN  20198.8  39535.0  

[1113 rows x 230 columns]

In [15]:
df_peptides

patient_id  visit_month  visit_id  AADDTWEPFASGK  \
0             55            0      55_0      8984260.0   
1             55            6      55_6      8279770.0   
2             55           12     55_12      8382390.0   
3             55           36     55_36     10671500.0   
4            942            6     942_6      6177730.0   
...          ...          ...       ...            ...   
1108       64674           84  64674_84      7083630.0   
1109       65043            0   65043_0      7818630.0   
1110       65043           12  65043_12      8070390.0   
1111       65043           24  65043_24      7608150.0   
1112       65043           48  65043_48      7187220.0   

      AAFGQGSGPIMLDEVQC(UniMod_4)TGTEASLADC(UniMod_4)K  \
0                                              53855.6   
1                                              45251.9   
2                                              53000.9   
3                                              58108.4   
4                                              42682.6   
...                                                ...   
1108                                           35656.1   
1109                                           95033.0   
1110                                           76532.7   
1111                                           75401.6   
1112                                           72262.2   

      AAFTEC(UniMod_4)C(UniMod_4)QAADK  AANEVSSADVK  AATGEC(UniMod_4)TATVGKR  \
0                            8579740.0          NaN                  19735.4   
1                            8655890.0      49927.5                  23820.4   
2                            8995640.0      45519.2                  17813.5   
3                            9985420.0      52374.0                  19373.3   
4                            3596660.0      25698.8                  17130.6   
...                                ...          ...                      ...   
1108                         6273100.0          NaN                      NaN   
1109                         5119260.0      57483.7                  11610.0   
1110                         8233520.0      54260.6                  11631.9   
1111                         9168030.0          NaN                  13313.9   
1112                         8551210.0          NaN                  12365.3   

      AATVGSLAGQPLQER  AAVYHHFISDGVR  ...  YSLTYIYTGLSK   YTTEIIK  \
0            114400.0        46371.1  ...      201158.0  16492.30   
1             90539.4        38652.4  ...      171079.0  13198.80   
2            147312.0        45840.9  ...      231772.0  17873.80   
3             64356.1        49793.2  ...      185290.0  18580.50   
4             86471.5        41007.9  ...      226314.0   6399.80   
...               ...            ...  ...           ...       ...   
1108          15479.2            NaN  ...      203523.0   3835.58   
1109         270739.0        42527.3  ...      257361.0  18316.60   
1110         230169.0        42255.5  ...      230437.0  16703.20   
1111         220202.0        46914.1  ...      251228.0  18326.20   
1112         197981.0        42311.3  ...      233567.0  14478.30   

      YVGGQEHFAHLLILR  YVM(UniMod_35)LPVADQDQC(UniMod_4)IR  \
0           3810270.0                             106894.0   
1           4119520.0                             113385.0   
2           5474140.0                             116286.0   
3           2659660.0                              90936.9   
4                 NaN                              57571.4   
...               ...                                  ...   
1108        4901220.0                              40325.9   
1109        2514660.0                              51444.6   
1110        2481560.0                              44405.0   
1111        2939460.0                              50588.2   
1112        3185530.0                              48793.0   

      YVMLPVADQDQC(UniMod_4)IR  YVNKEIQNAVNGVK  YWGVASFLQK  \
0              

Because not all 1000+ proteins and peptides are measured at each recorded visit, some NaN values should be now expected. 

In [16]:
df_proteins.isna().sum().sort_values(ascending=False)

Q99829        624
Q99832        507
Q562R1        497
P01780        459
Q6UX71        452
             ... 
P02766          0
P02765          0
P02751          0
P02749          0
patient_id      0
Length: 230, dtype: int64

In [17]:
df_peptides.isna().sum().sort_values(ascending=False)

QALPQVR                   624
EPQVYTLPPSRDELTK          550
TPSGLYLGTC(UniMod_4)ER    523
SLEDQVEMLR                514
HYEGSTVPEK                508
                         ... 
visit_id                    0
IPTTFENGR                   0
AIGYLNTGYQR                 0
NILTSNNIDVK                 0
patient_id                  0
Length: 971, dtype: int64

We are going to combine the protein and peptide data, and check once again the status of NaN values. They should remain unchanged because we haven't done anything with them.

In [18]:
prot_pept_df = pd.merge(df_proteins, df_peptides, on=['patient_id','visit_month','visit_id'], how='left')
prot_pept_df

patient_id  visit_month  visit_id   O00391     O00533   O00584   O14498  \
0             55            0      55_0  11254.3   732430.0  39585.8  41526.9   
1             55            6      55_6  13163.6   630465.0  35220.8  41295.0   
2             55           12     55_12  15257.6   815083.0  41650.9  39763.3   
3             55           36     55_36  13530.8   753832.0  43048.9  43503.6   
4            942            6     942_6  11218.7   399518.0  20581.0  31290.9   
...          ...          ...       ...      ...        ...      ...      ...   
1108       64674           84  64674_84      NaN   190487.0  24907.9  18543.1   
1109       65043            0   65043_0  13472.4   927954.0  42661.5  43663.2   
1110       65043           12  65043_12  14134.9   984651.0  28990.8  42440.9   
1111       65043           24  65043_24  14659.5  1062020.0  46440.4  38293.0   
1112       65043           48  65043_48  10589.6   902434.0  44890.8  38771.5   

        O14773   O14791    O15240  ...  YSLTYIYTGLSK   YTTEIIK  \
0     31238.00  4202.71  177775.0  ...      201158.0  16492.30   
1     26219.90  4416.42  165638.0  ...      171079.0  13198.80   
2     30703.60  4343.60  151073.0  ...      231772.0  17873.80   
3     33577.60  5367.06  101056.0  ...      185290.0  18580.50   
4      6173.58  2564.37  160526.0  ...      226314.0   6399.80   
...        ...      ...       ...  ...           ...       ...   
1108  10124.90  2308.71   62095.4  ...      203523.0   3835.58   
1109  20071.30  3278.88  266339.0  ...      257361.0  18316.60   
1110  25357.40  3267.66  270575.0  ...      230437.0  16703.20   
1111  21971.80  3990.34  221358.0  ...      251228.0  18326.20   
1112  20188.30  3137.07  203487.0  ...      233567.0  14478.30   

      YVGGQEHFAHLLILR  YVM(UniMod_35)LPVADQDQC(UniMod_4)IR  \
0           3810270.0                             106894.0   
1           4119520.0                             113385.0   
2           5474140.0                             116286.0   
3           2659660.0                              90936.9   
4                 NaN                              57571.4   
...               ...                                  ...   
1108        4901220.0                              40325.9   
1109        2514660.0                              51444.6   
1110        2481560.0                              44405.0   
1111        2939460.0                              50588.2   
1112        3185530.0                              48793.0   

      YVMLPVADQDQC(UniMod_4)IR  YVNKEIQNAVNGVK  YWGVASFLQK  \
0                     580667.0        131155.0    165851.0   
1                     514861.0        103512.0    144607.0   
2                     711815.0        136943.0    181763.0   
3                     679163.0        128593.0    203680.0   
4                     480951.0         80001.2     79661.9   
...                        ...             ...         ...   
1108                  335625.0         49250.4     64076.3   
1109                  530245.0        156148.0    157548.0   
1110                  543391.0        159828.0    161207.0   
1111                  597869.0        148032.0    192857.0   
1112                  501159.0        133992.0    170146.0   

      YYC(UniMod_4)FQGNQFLR  YYTYLIMNK  YYWGGQYTWDMAK  
0                  437305.0    46289.2        14898.4  
1                  457891.0    40047.7        20703.9  
2                  452253.0    54725.1        21841.1  
3                  498621.0    52792.7        13973.7  
4                  573300.0    48005.8        15674.1  
...                     ...        ...            ...  
1108               667993.0    38472.5        21949.1  
1109               336625.0    48423.2        10915.8  
1110               330337.0    45368.1        19023.2  
1111               388125.0    65101.0        20790.1  
1112               359045.0    45780.0        17370.6  

[1113 rows x 1198 columns]

In [19]:
prot_pept_df.isna().sum().sort_values(ascending=False)

Q99829                    624
QALPQVR                   624
EPQVYTLPPSRDELTK          550
TPSGLYLGTC(UniMod_4)ER    523
SLEDQVEMLR                514
                         ... 
P41222                      0
P02774                      0
P02787                      0
P02790                      0
patient_id                  0
Length: 1198, dtype: int64

In [20]:
patient_list=prot_pept_df['patient_id'].unique()
patient_list

array([   55,   942,  1517,  1923,  2660,  3636,  3863,  4161,  4172,
        4923,  5027,  5036,  5178,  5645,  5742,  6054,  6211,  6420,
        7051,  7117,  7151,  7265,  7508,  7568,  7832,  7886,  8344,
        8699, 10053, 10138, 10174, 10541, 10715, 10718, 11459, 11686,
       11928, 12516, 12636, 12703, 12755, 12931, 13360, 13368, 13618,
       13804, 13852, 13968, 14035, 14124, 14242, 14270, 14344, 14450,
       14811, 15009, 15245, 15504, 15590, 16238, 16347, 16566, 16574,
       16778, 16931, 17154, 17201, 17414, 17727, 18183, 18204, 18553,
       18560, 19088, 20212, 20216, 20352, 20404, 20460, 20581, 20664,
       20707, 20791, 20792, 21126, 21537, 21729, 22126, 22623, 23175,
       23192, 23244, 23391, 23636, 24278, 24690, 24818, 24820, 24911,
       25562, 25739, 25750, 25827, 25911, 26005, 26104, 26210, 26809,
       27079, 27300, 27464, 27468, 27607, 27715, 27872, 27893, 27971,
       27987, 28327, 28342, 28818, 29313, 29417, 30119, 30155, 30416,
       30894, 30951,

Here there are two ways to impute the data:
1) combine "clinical" and "prot_pept_df" first then impute, or
2) first impute them separately, then combine. 

We'll do both and see how the outcomes differ.

I am choosing to use sklearn's KNN imputer because, with a bit of clever coding, I can use all the availble data for a given patient to impute missing values. 

In [21]:
# Method 1: Combine first, impute next. 
big_data = pd.merge(clinical, prot_pept_df, on=['patient_id','visit_month','visit_id'], how='left')
big_data.head(30)

visit_id  patient_id  visit_month  updrs_1  updrs_2  updrs_3  updrs_4  \
0      55_0          55            0     10.0      6.0     15.0      NaN   
1      55_3          55            3     10.0      7.0     25.0      NaN   
2      55_6          55            6      8.0     10.0     34.0      NaN   
3      55_9          55            9      8.0      9.0     30.0      0.0   
4     55_12          55           12     10.0     10.0     41.0      0.0   
5     55_18          55           18      7.0     13.0     38.0      0.0   
6     55_24          55           24     16.0      9.0     49.0      0.0   
7     55_30          55           30     14.0     13.0     49.0      0.0   
8     55_36          55           36     17.0     18.0     51.0      0.0   
9     55_42          55           42     12.0     20.0     41.0      0.0   
10    55_48          55           48     17.0     16.0     52.0      0.0   
11    55_54          55           54     12.0     18.0     51.0      0.0   
12    55_60          55           60     23.0     21.0     56.0      0.0   
13    942_0         942            0      3.0      2.0     20.0      NaN   
14    942_3         942            3      7.0      2.0     17.0      NaN   
15    942_6         942            6      8.0      2.0     21.0      NaN   
16   942_12         942           12      5.0      2.0     25.0      0.0   
17   942_18         942           18      6.0      5.0     18.0      0.0   
18   942_24         942           24      2.0      3.0     23.0      NaN   
19   942_30         942           30      4.0      4.0     19.0      0.0   
20   942_36         942           36      2.0      4.0     19.0      0.0   
21   942_42         942           42      2.0      3.0     32.0      0.0   
22   942_48         942           48      2.0      6.0     35.0      0.0   
23   942_54         942           54      2.0      4.0     26.0      0.0   
24   942_60         942           60      2.0      5.0     31.0      0.0   
25   942_72         942           72      4.0      2.0     33.0      0.0   
26   942_84         942           84      4.0      8.0     26.0      0.0   
27   942_96         942           96      6.0      9.0     39.0      3.0   
28   1517_0        1517            0     11.0      6.0     25.0      5.0   
29   1517_6        1517            6     17.0      4.0     30.0      7.0   

      O00391    O00533   O00584  ...  YSLTYIYTGLSK   YTTEIIK  YVGGQEHFAHLLILR  \
0   11254.30  732430.0  39585.8  ...      201158.0  16492.30        3810270.0   
1        NaN       NaN      NaN  ...           NaN       NaN              NaN   
2   13163.60  630465.0  35220.8  ...      171079.0  13198.80        4119520.0   
3        NaN       NaN      NaN  ...           NaN       NaN              NaN   
4   15257.60  815083.0  41650.9  ...      231772.0  17873.80        5474140.0   
5        NaN       NaN      NaN  ...           NaN       NaN              NaN   
6        NaN       NaN      NaN  ...           NaN       NaN              NaN   
7        NaN       NaN      NaN  ...           NaN       NaN              NaN   
8   13530.80  753832.0  43048.9  ...      185290.0  18580.50        2659660.0   
9        NaN       NaN      NaN  ...           NaN       NaN              NaN   
10       NaN       NaN      NaN  ...           NaN       NaN              NaN   
11       NaN       NaN      NaN  ...           NaN       NaN              NaN   
12       NaN       NaN      NaN  ...           NaN       NaN              NaN   
13       NaN       NaN      NaN  ...           NaN       NaN              NaN   
14       NaN       NaN      NaN  ...           NaN       NaN              NaN   
15  11218.70  399518.0  20581.0  ...      226314.0   6399.80              NaN   
16   6757.32  360858.0  18367.6  ...      173259.0   4767.63         374307.0   
17       NaN       NaN      NaN  ...           NaN       NaN              NaN   
18       NaN  352722.0  22834.9  ...      185428.0   5554.53              NaN   
19       NaN       NaN    

In [22]:
#imputing
big_data_patient_list = big_data.patient_id.unique()
data_imputed_list = []
for patient_id in big_data_patient_list:
    masked_data = big_data[big_data['patient_id']==patient_id]
    num_rows = len(masked_data.index)
    knn = KNNImputer(missing_values=np.nan, keep_empty_features=True, n_neighbors=num_rows)
    X_knn = knn.fit_transform(masked_data)
    X_knn_df = pd.DataFrame(X_knn, columns = big_data.columns)
    data_imputed_list.append(X_knn_df)

big_data_imputed = pd.concat(data_imputed_list)

In [23]:
big_data_imputed.head(10)

visit_id  patient_id  visit_month  updrs_1  updrs_2  updrs_3  updrs_4  \
0     550.0        55.0          0.0     10.0      6.0     15.0      0.0   
1     553.0        55.0          3.0     10.0      7.0     25.0      0.0   
2     556.0        55.0          6.0      8.0     10.0     34.0      0.0   
3     559.0        55.0          9.0      8.0      9.0     30.0      0.0   
4    5512.0        55.0         12.0     10.0     10.0     41.0      0.0   
5    5518.0        55.0         18.0      7.0     13.0     38.0      0.0   
6    5524.0        55.0         24.0     16.0      9.0     49.0      0.0   
7    5530.0        55.0         30.0     14.0     13.0     49.0      0.0   
8    5536.0        55.0         36.0     17.0     18.0     51.0      0.0   
9    5542.0        55.0         42.0     12.0     20.0     41.0      0.0   

      O00391    O00533   O00584  ...  YSLTYIYTGLSK   YTTEIIK  YVGGQEHFAHLLILR  \
0  11254.300  732430.0  39585.8  ...     201158.00  16492.30        3810270.0   
1  13301.575  732952.5  39876.6  ...     197324.75  16536.35        4015897.5   
2  13163.600  630465.0  35220.8  ...     171079.00  13198.80        4119520.0   
3  13301.575  732952.5  39876.6  ...     197324.75  16536.35        4015897.5   
4  15257.600  815083.0  41650.9  ...     231772.00  17873.80        5474140.0   
5  13301.575  732952.5  39876.6  ...     197324.75  16536.35        4015897.5   
6  13301.575  732952.5  39876.6  ...     197324.75  16536.35        4015897.5   
7  13301.575  732952.5  39876.6  ...     197324.75  16536.35        4015897.5   
8  13530.800  753832.0  43048.9  ...     185290.00  18580.50        2659660.0   
9  13301.575  732952.5  39876.6  ...     197324.75  16536.35        4015897.5   

   YVM(UniMod_35)LPVADQDQC(UniMod_4)IR  YVMLPVADQDQC(UniMod_4)IR  \
0                           106894.000                  580667.0   
1                           106875.475                  621626.5   
2                           113385.000                  514861.0   
3                           106875.475                  621626.5   
4                           116286.000                  711815.0   
5                           106875.475                  621626.5   
6                           106875.475                  621626.5   
7                           106875.475                  621626.5   
8                            90936.900                  679163.0   
9                           106875.475                  621626.5   

   YVNKEIQNAVNGVK  YWGVASFLQK  YYC(UniMod_4)FQGNQFLR  YYTYLIMNK  YYWGGQYTWDMAK  
0       131155.00   165851.00               437305.0  46289.200      14898.400  
1       125050.75   173975.25               461517.5  48463.675      17854.275  
2       103512.00   144607.00               457891.0  40047.700      20703.900  
3       125050.75   173975.25               461517.5  48463.675      17854.275  
4       136943.00   181763.00               452253.0  54725.100      21841.100  
5       125050.75   173975.25               461517.5  48463.675      17854.275  
6       125050.75   173975.25               461517.5  48463.675      17854.275  
7       125050.75   173975.25               461517.5  48463.675      17854.275  
8       128593.00   203680.00               498621.0  52792.700      13973.700  
9       125050.75   173975.25               461517.5  48463.675      17854.275  

[10 rows x 1202 columns]

In [24]:
#resetting index to match original.
big_data_imputed.reset_index(inplace=True, drop=True)
big_data_imputed.head(10)

visit_id  patient_id  visit_month  updrs_1  updrs_2  updrs_3  updrs_4  \
0     550.0        55.0          0.0     10.0      6.0     15.0      0.0   
1     553.0        55.0          3.0     10.0      7.0     25.0      0.0   
2     556.0        55.0          6.0      8.0     10.0     34.0      0.0   
3     559.0        55.0          9.0      8.0      9.0     30.0      0.0   
4    5512.0        55.0         12.0     10.0     10.0     41.0      0.0   
5    5518.0        55.0         18.0      7.0     13.0     38.0      0.0   
6    5524.0        55.0         24.0     16.0      9.0     49.0      0.0   
7    5530.0        55.0         30.0     14.0     13.0     49.0      0.0   
8    5536.0        55.0         36.0     17.0     18.0     51.0      0.0   
9    5542.0        55.0         42.0     12.0     20.0     41.0      0.0   

      O00391    O00533   O00584  ...  YSLTYIYTGLSK   YTTEIIK  YVGGQEHFAHLLILR  \
0  11254.300  732430.0  39585.8  ...     201158.00  16492.30        3810270.0   
1  13301.575  732952.5  39876.6  ...     197324.75  16536.35        4015897.5   
2  13163.600  630465.0  35220.8  ...     171079.00  13198.80        4119520.0   
3  13301.575  732952.5  39876.6  ...     197324.75  16536.35        4015897.5   
4  15257.600  815083.0  41650.9  ...     231772.00  17873.80        5474140.0   
5  13301.575  732952.5  39876.6  ...     197324.75  16536.35        4015897.5   
6  13301.575  732952.5  39876.6  ...     197324.75  16536.35        4015897.5   
7  13301.575  732952.5  39876.6  ...     197324.75  16536.35        4015897.5   
8  13530.800  753832.0  43048.9  ...     185290.00  18580.50        2659660.0   
9  13301.575  732952.5  39876.6  ...     197324.75  16536.35        4015897.5   

   YVM(UniMod_35)LPVADQDQC(UniMod_4)IR  YVMLPVADQDQC(UniMod_4)IR  \
0                           106894.000                  580667.0   
1                           106875.475                  621626.5   
2                           113385.000                  514861.0   
3                           106875.475                  621626.5   
4                           116286.000                  711815.0   
5                           106875.475                  621626.5   
6                           106875.475                  621626.5   
7                           106875.475                  621626.5   
8                            90936.900                  679163.0   
9                           106875.475                  621626.5   

   YVNKEIQNAVNGVK  YWGVASFLQK  YYC(UniMod_4)FQGNQFLR  YYTYLIMNK  YYWGGQYTWDMAK  
0       131155.00   165851.00               437305.0  46289.200      14898.400  
1       125050.75   173975.25               461517.5  48463.675      17854.275  
2       103512.00   144607.00               457891.0  40047.700      20703.900  
3       125050.75   173975.25               461517.5  48463.675      17854.275  
4       136943.00   181763.00               452253.0  54725.100      21841.100  
5       125050.75   173975.25               461517.5  48463.675      17854.275  
6       125050.75   173975.25               461517.5  48463.675      17854.275  
7       125050.75   173975.25               461517.5  48463.675      17854.275  
8       128593.00   203680.00               498621.0  52792.700      13973.700  
9       125050.75   173975.25               461517.5  48463.675      17854.275  

[10 rows x 1202 columns]

In [25]:
#relabel patient id, visit id, and visit month.
ls = [ids, month]
for col in ls:
    big_data_imputed[col] = big_data[col]

In [26]:
big_data_imputed.head(30)


visit_id  patient_id  visit_month  updrs_1  updrs_2  updrs_3   updrs_4  \
0      55_0          55            0     10.0      6.0     15.0  0.000000   
1      55_3          55            3     10.0      7.0     25.0  0.000000   
2      55_6          55            6      8.0     10.0     34.0  0.000000   
3      55_9          55            9      8.0      9.0     30.0  0.000000   
4     55_12          55           12     10.0     10.0     41.0  0.000000   
5     55_18          55           18      7.0     13.0     38.0  0.000000   
6     55_24          55           24     16.0      9.0     49.0  0.000000   
7     55_30          55           30     14.0     13.0     49.0  0.000000   
8     55_36          55           36     17.0     18.0     51.0  0.000000   
9     55_42          55           42     12.0     20.0     41.0  0.000000   
10    55_48          55           48     17.0     16.0     52.0  0.000000   
11    55_54          55           54     12.0     18.0     51.0  0.000000   
12    55_60          55           60     23.0     21.0     56.0  0.000000   
13    942_0         942            0      3.0      2.0     20.0  0.272727   
14    942_3         942            3      7.0      2.0     17.0  0.272727   
15    942_6         942            6      8.0      2.0     21.0  0.272727   
16   942_12         942           12      5.0      2.0     25.0  0.000000   
17   942_18         942           18      6.0      5.0     18.0  0.000000   
18   942_24         942           24      2.0      3.0     23.0  0.272727   
19   942_30         942           30      4.0      4.0     19.0  0.000000   
20   942_36         942           36      2.0      4.0     19.0  0.000000   
21   942_42         942           42      2.0      3.0     32.0  0.000000   
22   942_48         942           48      2.0      6.0     35.0  0.000000   
23   942_54         942           54      2.0      4.0     26.0  0.000000   
24   942_60         942           60      2.0      5.0     31.0  0.000000   
25   942_72         942           72      4.0      2.0     33.0  0.000000   
26   942_84         942           84      4.0      8.0     26.0  0.000000   
27   942_96         942           96      6.0      9.0     39.0  3.000000   
28   1517_0        1517            0     11.0      6.0     25.0  5.000000   
29   1517_6        1517            6     17.0      4.0     30.0  7.000000   

       O00391    O00533   O00584  ...  YSLTYIYTGLSK     YTTEIIK  \
0   11254.300  732430.0  39585.8  ...     201158.00  16492.3000   
1   13301.575  732952.5  39876.6  ...     197324.75  16536.3500   
2   13163.600  630465.0  35220.8  ...     171079.00  13198.8000   
3   13301.575  732952.5  39876.6  ...     197324.75  16536.3500   
4   15257.600  815083.0  41650.9  ...     231772.00  17873.8000   
5   13301.575  732952.5  39876.6  ...     197324.75  16536.3500   
6   13301.575  732952.5  39876.6  ...     197324.75  16536.3500   
7   13301.575  732952.5  39876.6  ...     197324.75  16536.3500   
8   13530.800  753832.0  43048.9  ...     185290.00  18580.5000   
9   13301.575  732952.5  39876.6  ...     197324.75  16536.3500   
10  13301.575  732952.5  39876.6  ...     197324.75  16536.3500   
11  13301.575  732952.5  39876.6  ...     197324.75  16536.3500   
12  13301.575  732952.5  39876.6  ...     197324.75  16536.3500   
13   9867.940  341229.5  20957.5  ...     180653.00   5758.0125   
14   9867.940  341229.5  20957.5  ...     180653.00   5758.0125   
15  11218.700  399518.0  20581.0  ...     226314.00   6399.8000   
16   6757.320  360858.0  18367.6  ...     173259.00   4767.6300   
17   9867.940  341229.5  20957.5  ...     180653.00   5758.0125   
18   9867.940  352722.0  22834.9  ...     185428.00   5554.5300   
19   9867.940  341229.5  20957.5  ...     180653.00   5758.0125   
20   9867.940  341229.5  20957.5  ...     180653.00   5758.0125   
21   9867.940  341229.5  20957.5  ...     180653.00   5758.0125   
22  11627.800  251820.0  22046.5  ...     137611.00   6310.0900   
23   98

Next is Method #2: impute each dataset first, then combine. 

Imputing prot_pept_df:

In [27]:
data_imputed_list = []
for patient_id in patient_list:
    masked_data = prot_pept_df[prot_pept_df['patient_id']==patient_id]
    num_rows = len(masked_data.index)
    knn = KNNImputer(missing_values=np.nan, keep_empty_features=True, n_neighbors=num_rows)
    X_knn = knn.fit_transform(masked_data)
    X_knn_df = pd.DataFrame(X_knn, columns = prot_pept_df.columns)
    data_imputed_list.append(X_knn_df)
prot_pept_imputed = pd.concat(data_imputed_list)

In [28]:
prot_pept_imputed.isna().sum()

patient_id               0
visit_month              0
visit_id                 0
O00391                   0
O00533                   0
                        ..
YVNKEIQNAVNGVK           0
YWGVASFLQK               0
YYC(UniMod_4)FQGNQFLR    0
YYTYLIMNK                0
YYWGGQYTWDMAK            0
Length: 1198, dtype: int64

In [29]:
prot_pept_imputed.reset_index(inplace=True, drop=True)
prot_pept_imputed

patient_id  visit_month   visit_id   O00391     O00533   O00584  \
0           55.0          0.0      550.0  11254.3   732430.0  39585.8   
1           55.0          6.0      556.0  13163.6   630465.0  35220.8   
2           55.0         12.0     5512.0  15257.6   815083.0  41650.9   
3           55.0         36.0     5536.0  13530.8   753832.0  43048.9   
4          942.0          6.0     9426.0  11218.7   399518.0  20581.0   
...          ...          ...        ...      ...        ...      ...   
1108     64674.0         84.0  6467484.0      0.0   190487.0  24907.9   
1109     65043.0          0.0   650430.0  13472.4   927954.0  42661.5   
1110     65043.0         12.0  6504312.0  14134.9   984651.0  28990.8   
1111     65043.0         24.0  6504324.0  14659.5  1062020.0  46440.4   
1112     65043.0         48.0  6504348.0  10589.6   902434.0  44890.8   

       O14498    O14773   O14791    O15240  ...  YSLTYIYTGLSK   YTTEIIK  \
0     41526.9  31238.00  4202.71  177775.0  ...      201158.0  16492.30   
1     41295.0  26219.90  4416.42  165638.0  ...      171079.0  13198.80   
2     39763.3  30703.60  4343.60  151073.0  ...      231772.0  17873.80   
3     43503.6  33577.60  5367.06  101056.0  ...      185290.0  18580.50   
4     31290.9   6173.58  2564.37  160526.0  ...      226314.0   6399.80   
...       ...       ...      ...       ...  ...           ...       ...   
1108  18543.1  10124.90  2308.71   62095.4  ...      203523.0   3835.58   
1109  43663.2  20071.30  3278.88  266339.0  ...      257361.0  18316.60   
1110  42440.9  25357.40  3267.66  270575.0  ...      230437.0  16703.20   
1111  38293.0  21971.80  3990.34  221358.0  ...      251228.0  18326.20   
1112  38771.5  20188.30  3137.07  203487.0  ...      233567.0  14478.30   

      YVGGQEHFAHLLILR  YVM(UniMod_35)LPVADQDQC(UniMod_4)IR  \
0           3810270.0                             106894.0   
1           4119520.0                             113385.0   
2           5474140.0                             116286.0   
3           2659660.0                              90936.9   
4            374307.0                              57571.4   
...               ...                                  ...   
1108        4901220.0                              40325.9   
1109        2514660.0                              51444.6   
1110        2481560.0                              44405.0   
1111        2939460.0                              50588.2   
1112        3185530.0                              48793.0   

      YVMLPVADQDQC(UniMod_4)IR  YVNKEIQNAVNGVK  YWGVASFLQK  \
0                     580667.0        131155.0    165851.0   
1                     514861.0        103512.0    144607.0   
2                     711815.0        136943.0    181763.0   
3                     679163.0        128593.0    203680.0   
4                     480951.0         80001.2     79661.9   
...                        ...             ...         ...   
1108                  335625.0         49250.4     64076.3   
1109                  530245.0        156148.0    157548.0   
1110                  543391.0        159828.0    161207.0   
1111                  597869.0        148032.0    192857.0   
1112                  501159.0        133992.0    170146.0   

      YYC(UniMod_4)FQGNQFLR  YYTYLIMNK  YYWGGQYTWDMAK  
0                  437305.0    46289.2        14898.4  
1                  457891.0    40047.7        20703.9  
2                  452253.0    54725.1        21841.1  
3                  498621.0    52792.7        13973.7  
4                  573300.0    48005.8        15674.1  
...                     ...        ...            ...  
1108               667993.0    38472.5        21949.1  
1109               336625.0    48423.2        10915.8  
1110               330337.0    45368.1        19023.2  
1111               388125.0    65101.0        20790.1  
1112               359045.0    45780.0        17370.6  

[1113 rows x 1198 columns]

In [30]:
prot_pept_imputed = prot_pept_imputed.astype({'patient_id': 'int', 'visit_month': 'int'})
prot_pept_imputed['visit_id'] = prot_pept_df['visit_id']
print(prot_pept_imputed.dtypes)

patient_id                 int32
visit_month                int32
visit_id                  object
O00391                   float64
O00533                   float64
                          ...   
YVNKEIQNAVNGVK           float64
YWGVASFLQK               float64
YYC(UniMod_4)FQGNQFLR    float64
YYTYLIMNK                float64
YYWGGQYTWDMAK            float64
Length: 1198, dtype: object


In [31]:
prot_pept_imputed

patient_id  visit_month  visit_id   O00391     O00533   O00584   O14498  \
0             55            0      55_0  11254.3   732430.0  39585.8  41526.9   
1             55            6      55_6  13163.6   630465.0  35220.8  41295.0   
2             55           12     55_12  15257.6   815083.0  41650.9  39763.3   
3             55           36     55_36  13530.8   753832.0  43048.9  43503.6   
4            942            6     942_6  11218.7   399518.0  20581.0  31290.9   
...          ...          ...       ...      ...        ...      ...      ...   
1108       64674           84  64674_84      0.0   190487.0  24907.9  18543.1   
1109       65043            0   65043_0  13472.4   927954.0  42661.5  43663.2   
1110       65043           12  65043_12  14134.9   984651.0  28990.8  42440.9   
1111       65043           24  65043_24  14659.5  1062020.0  46440.4  38293.0   
1112       65043           48  65043_48  10589.6   902434.0  44890.8  38771.5   

        O14773   O14791    O15240  ...  YSLTYIYTGLSK   YTTEIIK  \
0     31238.00  4202.71  177775.0  ...      201158.0  16492.30   
1     26219.90  4416.42  165638.0  ...      171079.0  13198.80   
2     30703.60  4343.60  151073.0  ...      231772.0  17873.80   
3     33577.60  5367.06  101056.0  ...      185290.0  18580.50   
4      6173.58  2564.37  160526.0  ...      226314.0   6399.80   
...        ...      ...       ...  ...           ...       ...   
1108  10124.90  2308.71   62095.4  ...      203523.0   3835.58   
1109  20071.30  3278.88  266339.0  ...      257361.0  18316.60   
1110  25357.40  3267.66  270575.0  ...      230437.0  16703.20   
1111  21971.80  3990.34  221358.0  ...      251228.0  18326.20   
1112  20188.30  3137.07  203487.0  ...      233567.0  14478.30   

      YVGGQEHFAHLLILR  YVM(UniMod_35)LPVADQDQC(UniMod_4)IR  \
0           3810270.0                             106894.0   
1           4119520.0                             113385.0   
2           5474140.0                             116286.0   
3           2659660.0                              90936.9   
4            374307.0                              57571.4   
...               ...                                  ...   
1108        4901220.0                              40325.9   
1109        2514660.0                              51444.6   
1110        2481560.0                              44405.0   
1111        2939460.0                              50588.2   
1112        3185530.0                              48793.0   

      YVMLPVADQDQC(UniMod_4)IR  YVNKEIQNAVNGVK  YWGVASFLQK  \
0                     580667.0        131155.0    165851.0   
1                     514861.0        103512.0    144607.0   
2                     711815.0        136943.0    181763.0   
3                     679163.0        128593.0    203680.0   
4                     480951.0         80001.2     79661.9   
...                        ...             ...         ...   
1108                  335625.0         49250.4     64076.3   
1109                  530245.0        156148.0    157548.0   
1110                  543391.0        159828.0    161207.0   
1111                  597869.0        148032.0    192857.0   
1112                  501159.0        133992.0    170146.0   

      YYC(UniMod_4)FQGNQFLR  YYTYLIMNK  YYWGGQYTWDMAK  
0                  437305.0    46289.2        14898.4  
1                  457891.0    40047.7        20703.9  
2                  452253.0    54725.1        21841.1  
3                  498621.0    52792.7        13973.7  
4                  573300.0    48005.8        15674.1  
...                     ...        ...            ...  
1108               667993.0    38472.5        21949.1  
1109               336625.0    48423.2        10915.8  
1110               330337.0    45368.1        19023.2  
1111               388125.0    65101.0        20790.1  
1112               359045.0    45780.0        17370.6  

[1113 rows x 1198 columns]

In [32]:
print(clinical.isna().sum())
print(f'dataset shape: {clinical.shape}')

visit_id          0
patient_id        0
visit_month       0
updrs_1           1
updrs_2           2
updrs_3          25
updrs_4        1038
dtype: int64
dataset shape: (2615, 7)


In [33]:
data_imputed_list = []
for patient_id in patient_list:
    masked_data = clinical[clinical['patient_id']==patient_id]
    num_rows = len(masked_data.index)
    knn = KNNImputer(missing_values=np.nan, keep_empty_features=True, n_neighbors=num_rows)
    X_knn = knn.fit_transform(masked_data)
    X_knn_df = pd.DataFrame(X_knn, columns = clinical.columns)
    data_imputed_list.append(X_knn_df)
clinical_imputed = pd.concat(data_imputed_list)

In [34]:
print(clinical_imputed.isna().sum())
print(f'imputed clinical dataset shape: {clinical_imputed.shape}')

visit_id       0
patient_id     0
visit_month    0
updrs_1        0
updrs_2        0
updrs_3        0
updrs_4        0
dtype: int64
imputed clinical dataset shape: (2615, 7)


In [35]:
clinical_imputed.reset_index(inplace=True, drop=True)
clinical_imputed

visit_id  patient_id  visit_month  updrs_1  updrs_2  updrs_3  updrs_4
0         550.0        55.0          0.0     10.0      6.0     15.0      0.0
1         553.0        55.0          3.0     10.0      7.0     25.0      0.0
2         556.0        55.0          6.0      8.0     10.0     34.0      0.0
3         559.0        55.0          9.0      8.0      9.0     30.0      0.0
4        5512.0        55.0         12.0     10.0     10.0     41.0      0.0
...         ...         ...          ...      ...      ...      ...      ...
2610  6504348.0     65043.0         48.0      7.0      6.0     13.0      0.0
2611  6504354.0     65043.0         54.0      4.0      8.0     11.0      1.0
2612  6504360.0     65043.0         60.0      6.0      6.0     16.0      1.0
2613  6504372.0     65043.0         72.0      3.0      9.0     14.0      1.0
2614  6504384.0     65043.0         84.0      7.0      9.0     20.0      3.0

[2615 rows x 7 columns]

In [36]:
clinical_imputed = clinical_imputed.astype({'patient_id': 'int', 'visit_month': 'int'})
clinical_imputed['visit_id'] = clinical['visit_id']
print(clinical_imputed.dtypes)

visit_id        object
patient_id       int32
visit_month      int32
updrs_1        float64
updrs_2        float64
updrs_3        float64
updrs_4        float64
dtype: object


In [37]:

prot_pept_clinical = pd.merge(prot_pept_imputed, clinical_imputed, on=['visit_id', 'visit_month', 'patient_id'], how='left')


In [38]:
print(prot_pept_clinical.isna().sum())
print(f'merged protein peptide and clinical dataset shape: {prot_pept_clinical.shape}')

patient_id        0
visit_month       0
visit_id          0
O00391            0
O00533            0
                 ..
YYWGGQYTWDMAK     0
updrs_1          45
updrs_2          45
updrs_3          45
updrs_4          45
Length: 1202, dtype: int64
merged protein peptide and clinical dataset shape: (1113, 1202)


Interestingly, with Method #2 (first impute separately, then combine) we end up with some additional NaN values that need to be dealt with. 

In [39]:
df = prot_pept_clinical[prot_pept_clinical.columns[prot_pept_clinical.isna().any()]]

df1 = df[df.isna().any(axis=1)]

print(df1)
print()
print(f'Remaining NaN values is a dataframe of: {df1.shape}')

      updrs_1  updrs_2  updrs_3  updrs_4
16        NaN      NaN      NaN      NaN
47        NaN      NaN      NaN      NaN
50        NaN      NaN      NaN      NaN
57        NaN      NaN      NaN      NaN
92        NaN      NaN      NaN      NaN
95        NaN      NaN      NaN      NaN
156       NaN      NaN      NaN      NaN
172       NaN      NaN      NaN      NaN
182       NaN      NaN      NaN      NaN
216       NaN      NaN      NaN      NaN
219       NaN      NaN      NaN      NaN
239       NaN      NaN      NaN      NaN
269       NaN      NaN      NaN      NaN
338       NaN      NaN      NaN      NaN
342       NaN      NaN      NaN      NaN
364       NaN      NaN      NaN      NaN
378       NaN      NaN      NaN      NaN
405       NaN      NaN      NaN      NaN
406       NaN      NaN      NaN      NaN
411       NaN      NaN      NaN      NaN
431       NaN      NaN      NaN      NaN
439       NaN      NaN      NaN      NaN
474       NaN      NaN      NaN      NaN
499       NaN   

In this case, all four parts of the UPDRS score are empty. The best course of action would be to drop the 45 rows entirely, since there is no reasonable way to impute the NaN values (or else we wouldn't need to train a model to predict these values).

In [40]:
prot_pept_clinical.dropna(subset = targets, inplace=True, axis=0)
prot_pept_clinical.isna().sum()

patient_id       0
visit_month      0
visit_id         0
O00391           0
O00533           0
                ..
YYWGGQYTWDMAK    0
updrs_1          0
updrs_2          0
updrs_3          0
updrs_4          0
Length: 1202, dtype: int64

Ok, so now we have two imputed combined datasets to work with. Before introducing ML models, I'll first define a function to calculate SMAPE scores. 

In [41]:
def smape_score(actual, predicted):
    sum = 0
    for a, p in zip(actual, predicted):
        if a==0 and p==0:
            pass
        else:
            sum += (np.abs(p-a))/(np.abs(p)+np.abs(a))*2
    return sum/len(actual)*100

The models to be tested include:

Random Forest Regressor, HistGradientBoostingRegressor, Tree based AdaBoostRegressor, Tree based XGBRegressor.


Before specifying parameters for each model, my first question is: if I don't specify any parameters, what does the model do on its own? 

Let's first take a look at RandomForestRegressor training on big_data_imputed. 

In [161]:
RFR = RandomForestRegressor() # not specifying any parameters.

X = big_data_imputed.drop(columns=targets, axis=1)

for target in targets:

    y = big_data_imputed[target]
    
    #splitting training and testing data. 
    X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.9, random_state=0)

    print(f'{target}: {len(X_train)} samples in training, {len(X_test)} samples in testing.')
    
    #fitting and testing
    RFR.fit(X_train,y_train)    
    y_predict = RFR.predict(X_test)
    #scoring
    mae = mean_absolute_error(y_test, y_predict)
    smape = smape_score(y_test, y_predict)
    #saving the scores in respective lists
    print(f'MAE: {mae}; SMAPE: {smape}')
    print()

    max_depth = list()

    for tree in RFR.estimators_:
        max_depth.append(tree.tree_.max_depth)

    print("avg max depth %0.1f" % (sum(max_depth) / len(max_depth)))
    print(f"max depth {max(max_depth)}")
    print(f"min depth {min(max_depth)}")
    print()

updrs_1: 2353 samples in training, 262 samples in testing.
MAE: 2.3774777098085953; SMAPE: 48.9772989442121

avg max depth 27.8
max depth 32
min depth 20

updrs_2: 2353 samples in training, 262 samples in testing.
MAE: 2.310240627650551; SMAPE: 66.15554885563363

avg max depth 25.0
max depth 32
min depth 20

updrs_3: 2353 samples in training, 262 samples in testing.
MAE: 6.3805042779103855; SMAPE: 57.76029812084096

avg max depth 25.3
max depth 32
min depth 20

updrs_4: 2353 samples in training, 262 samples in testing.
MAE: 0.976007328361716; SMAPE: 140.45296525716998

avg max depth 30.7
max depth 32
min depth 21



In [167]:
depth, counts = np.unique(max_depth, return_counts=True)
print(f'unique depths: {depth}')
print(f'occurances: {counts}')

unique depths: [21 22 23 24 25 26 27 28 29 30 31 32]
occurances: [  1   2   3  10   8  13  20  26  31  33  35 318]


A random forest regressor with no specified parameters, training on this dataset, reaches a maximum depth of 32 in 318 (out of 500) trees, which means the models appears to favor deep trees - an important feature to keep in mind. 

For RandomForestRegressor, argurably the most important parameters to control are:

n_estimators, the number of decision trees in the forest;

max_depth, the max number of splits that each tree is allowed.

I'll select some parameters, based on feedback from the previous uncontrolled RandomForest Regressor, which noticeably favors deeper trees for this dataset. So I will set a range of max_depth from 8 to 60, and n_estimators (number of trees in the forest) from 100 (default) to 800. I'll then compare these models to the default model.

In [169]:

RFR = RandomForestRegressor()

params = {'n_estimators': [100,200,500,800],
                 'max_depth': [8,16,32,48,60]}

random_params = list(ParameterSampler(params, n_iter=10, random_state=0))

random_params


[{'n_estimators': 500, 'max_depth': 60},
 {'n_estimators': 200, 'max_depth': 8},
 {'n_estimators': 800, 'max_depth': 60},
 {'n_estimators': 100, 'max_depth': 32},
 {'n_estimators': 500, 'max_depth': 32},
 {'n_estimators': 200, 'max_depth': 60},
 {'n_estimators': 500, 'max_depth': 16},
 {'n_estimators': 200, 'max_depth': 48},
 {'n_estimators': 100, 'max_depth': 16},
 {'n_estimators': 500, 'max_depth': 8}]

I'll first save these pseudo randomly generated parameters in a dataframe, for later referencing.

In [170]:
params_df = pd.DataFrame(random_params)
params_df

n_estimators  max_depth
0           500         60
1           200          8
2           800         60
3           100         32
4           500         32
5           200         60
6           500         16
7           200         48
8           100         16
9           500          8

I'll first test the RandomForestRegressor model on "big_data_imputed", i.e. data that was first combined then imputed. 



In [175]:
#randomforestregressor 
smape_list_RFR1 = []
mae_list_RFR1 = []
actual_depth_RFR1 = []

X = big_data_imputed.drop(columns=targets, axis=1)

for target in targets:

    y = big_data_imputed[target]
    
    #splitting training and testing data. 
    X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.9, random_state=0)

    print(f'{target}: {len(X_train)} samples in training, {len(X_test)} samples in testing.')
    
    for params in random_params:
        #fitting and testing
        RFR.set_params(**params)
        RFR.fit(X_train,y_train)    
        y_predict = RFR.predict(X_test)
        #scoring
        mae = mean_absolute_error(y_test, y_predict)
        smape = smape_score(y_test, y_predict)
        #saving the scpres in respective lists
        mae_list_RFR1.append(mae)
        smape_list_RFR1.append(smape)
        print(f'Parameter: {params}; MAE: {mae}; SMAPE: {smape}')
        print()
        #still good to know how deep the trees are actually going.
        max_depth = list()
        for tree in RFR.estimators_:
            max_depth.append(tree.tree_.max_depth)
        print("avg max depth %0.1f" % (sum(max_depth) / len(max_depth)))
        print(f"max depth {max(max_depth)}")
        print(f"min depth {min(max_depth)}")
        actual_depth_RFR1.append(max(max_depth))
    print()


updrs_1: 2353 samples in training, 262 samples in testing.


I will store the scores and their respective parameters in a dataframe, for easy referencing. 

In [132]:
RFR_df = pd.concat([params_df,params_df,params_df,params_df], ignore_index=True)
RFR_df['Actual Max Depth'] = actual_depth_RFR1
RFR_df['MAE'] = mae_list_RFR1
RFR_df['SMAPE'] = smape_list_RFR1
RFR_df['UPDRS'] = ""
RFR_df.loc[:9, 'UPDRS']=1
RFR_df.loc[10:19, 'UPDRS']=2
RFR_df.loc[20:29, 'UPDRS']=3
RFR_df.loc[30:, 'UPDRS']=4

RFR_df

n_estimators  max_features  max_depth        MAE       SMAPE UPDRS
0            500            10          2   3.986906   64.208161     1
1            200           200          2   3.807458   62.530662     1
2            500           500          8   2.849360   52.574747     1
3            500            25          8   3.020394   54.119834     1
4            100           200         16   2.405778   48.390476     1
5            100           200          4   3.484876   59.251964     1
6            100           100          8   2.991610   54.007503     1
7            200            50          8   2.982202   53.591650     1
8            100            50         32   2.486042   50.464037     1
9            500           500         32   2.405551   49.523959     1
10           500            10          2   4.723631   86.435538     2
11           200           200          2   4.490907   84.926241     2
12           500           500          8   2.841651   72.225699     2
13           500            25          8   3.171161   75.300227     2
14           100           200         16   2.378240   68.925467     2
15           100           200          4   3.903892   80.411747     2
16           100           100          8   3.035141   73.908373     2
17           200            50          8   3.091225   74.376854     2
18           100            50         32   2.402313   62.428561     2
19           500           500         32   2.317888   66.285347     2
20           500            10          2  11.551000   74.922311     3
21           200           200          2  11.162200   73.683093     3
22           500           500          8   7.184333   61.297045     3
23           500            25          8   8.006649   64.651057     3
24           100           200         16   6.500928   58.610548     3
25           100           200          4   9.809870   69.930527     3
26           100           100          8   7.658528   63.335328     3
27           200            50          8   7.716575   63.344083     3
28           100            50         32   6.759921   59.014297     3
29           500           500         32   6.404088   57.751674     3
30           500            10          2   1.704855  150.427140     4
31           200           200          2   1.632735  149.553349     4
32           500           500          8   1.252739  145.868614     4
33           500            25          8   1.311646  145.355005     4
34           100           200         16   1.067220  142.808940     4
35           100           200          4   1.486425  148.113146     4
36           100           100          8   1.294648  145.705962     4
37           200            50          8   1.312905  145.449584     4
38           100            50         32   0.980778  136.632679     4
39           500           500         32   0.966471  140.436424     4

In [ ]:
a = RFR_df.style.highlight_min(axis=0, color='green', subset=['MAE', "SMAPE"][:9])

a

How to add a row of labels? updrs_1 to _4? and highlight the best MAE scores for each section?

I'll highlight the cells with the lowest repective MAE scores. Note that these may not be the lowest SMAPE scores, but I have already discussed and established that SMAPE may not be the best metric for this model. Mean Absolute Error, on the other hand, paints a clearer picture. ``

Note that the training dataset contains 2353 samples, and the testing dataset contains 262 samples. 

Now we'll test the same set of parameters on the same model, on "prot_pept_clinical" data, i.e. data that was first independently imputed, then combined. Like I have previously pointed out, this method produces a dataset that it less than 50% the size of Method 1. I would expect the trained model to perform more poorly. 

In [76]:
X = prot_pept_clinical.drop(columns=targets, axis=1)
smape_list_RFR2 = []
mae_list_RFR2 = []

for target in targets:

    y = prot_pept_clinical[target]
    
    #splitting training and testing data. 
    X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.9, random_state=0)

    print(f'{target}: {len(X_train)} samples in training, {len(X_test)} samples in testing.')
    
    for params in random_params:
        #fitting and testing
        RFR.set_params(**params)
        RFR.fit(X_train,y_train)    
        y_predict = RFR.predict(X_test)
        #scoring
        mae = mean_absolute_error(y_test, y_predict)
        smape = smape_score(y_test, y_predict)
        #saving the scpres in respective lists
        mae_list_RFR2.append(mae)
        smape_list_RFR2.append(smape)

        print(f'Parameter: {params}; MAE: {mae}; SMAPE: {smape}')

    print()

updrs_1: 961 samples in training, 107 samples in testing.
Parameter: {'n_estimators': 500, 'max_features': 10, 'max_depth': 2}; MAE: 3.416062050579234; SMAPE: 64.71977024918934
Parameter: {'n_estimators': 200, 'max_features': 200, 'max_depth': 2}; MAE: 3.2360592983851895; SMAPE: 62.86375365425435
Parameter: {'n_estimators': 500, 'max_features': 500, 'max_depth': 8}; MAE: 3.0005992441821463; SMAPE: 60.97880619277473
Parameter: {'n_estimators': 500, 'max_features': 25, 'max_depth': 8}; MAE: 3.1404803448998186; SMAPE: 62.01681521920457
Parameter: {'n_estimators': 100, 'max_features': 200, 'max_depth': 16}; MAE: 2.9911694017650303; SMAPE: 60.18451398621255
Parameter: {'n_estimators': 100, 'max_features': 200, 'max_depth': 4}; MAE: 3.115459506105469; SMAPE: 61.69806021009223
Parameter: {'n_estimators': 100, 'max_features': 100, 'max_depth': 8}; MAE: 3.0519897805945453; SMAPE: 61.3595290273705
Parameter: {'n_estimators': 200, 'max_features': 50, 'max_depth': 8}; MAE: 3.116670337356124; SMAPE

Overall, the model performs better across all tested parameters with the first dataset ("big_data_imputed"). This is as expected, since it contains more than 2 times the data as the other set. We will forgo the initial random parameter search with this smaller dataset on all the other models for now.

In [ ]:
#patient_id stratefied splitting
'''
smape_list = []
mae_list = []

for target in targets:
    X = big_data_imputed.drop(columns=targets, axis=1)
    y = big_data_imputed[target]

    for i, (train_index, test_index) in enumerate(gss.split(X, y, groups=X.patient_id)):
    
        X_train = X.loc[train_index]
        y_train = y.loc[train_index]
        X_test = X.loc[test_index]
        y_test = y.loc[test_index]
        
        print(f'{target} Split {i+1}: {len(X_train)} samples in training, {len(X_test)} samples in testing.')
    
        RFR.fit(X_train, y_train)

        y_predict = RFR.predict(X_test)

        mae = mean_absolute_error(y_test, y_predict)
        smape = smape_score(y_test, y_predict)

        smape_list.append(smape)
        mae_list.append(mae)

        print(f'MAE score for {target}: {mae}')
        print(f'SMAPE score for {target}: {smape}')
        print()

    mae_average = np.mean(mae_list)
    smape_average = np.mean(smape_list)

    print(f'MAE average score: {mae_average}')
    print(f'SMAPE average score: {smape_average}')
    print()

        

'''


HistGradientBoostingRegressor (HGBR) does not support max_features as a parameter, so we will leave it out here.


In [80]:
params_HGBR = {'min_samples_leaf': [3,5,10,15,20],
                 'max_depth': [2,4,8,16,32]}
random_params_HGBR = list(ParameterSampler(params_HGBR, n_iter=10, random_state=0))

#save them in a dataframe
params_df_hgb = pd.DataFrame(random_params_HGBR)
params_df_hgb

min_samples_leaf  max_depth
0                 3          4
1                10          2
2                20         16
3                 5         16
4                 5          8
5                10         32
6                10         16
7                20         32
8                15         32
9                20          8

In [144]:
hgb = HistGradientBoostingRegressor()

X = big_data_imputed.drop(columns=targets, axis=1)


for target in targets:

    y = big_data_imputed[target]
    
    #splitting training and testing data. 
    X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.9, random_state=10)

    print(f'{target}: {len(X_train)} samples in training, {len(X_test)} samples in testing.')

    #Randomly generate a dictionary containing 10 combinations of parameters. 
    
    #fitting and testing
    hgb.fit(X_train,y_train)    
    y_predict = hgb.predict(X_test)
    #scoring
    mae = mean_absolute_error(y_test, y_predict)
    smape = smape_score(y_test, y_predict)
    #saving scores in their lists
    print(f'MAE: {mae}; SMAPE: {smape}')

    print()

updrs_1: 2353 samples in training, 262 samples in testing.
MAE: 2.413906042706124; SMAPE: 52.37888245935983

updrs_2: 2353 samples in training, 262 samples in testing.
MAE: 2.2997074085595823; SMAPE: 69.2213090591029

updrs_3: 2353 samples in training, 262 samples in testing.
MAE: 5.971028618733929; SMAPE: 56.31828494260663

updrs_4: 2353 samples in training, 262 samples in testing.
MAE: 1.0867119467403688; SMAPE: 148.58266993117277



In [81]:
#HistGradientBoostingRegressor with big_data_imputed

hgb = HistGradientBoostingRegressor()

X = big_data_imputed.drop(columns=targets, axis=1)

smape_list_hgb = []
mae_list_hgb = []

for target in targets:

    y = big_data_imputed[target]
    
    #splitting training and testing data. 
    X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.9, random_state=0)

    print(f'{target}: {len(X_train)} samples in training, {len(X_test)} samples in testing.')

    #Randomly generate a dictionary containing 10 combinations of parameters. 
    
    for params in random_params_HGBR:
        #fitting and testing
        hgb.set_params(**params)
        hgb.fit(X_train,y_train)    
        y_predict = hgb.predict(X_test)
        #scoring
        mae = mean_absolute_error(y_test, y_predict)
        smape = smape_score(y_test, y_predict)
        #saving scores in their lists
        mae_list_hgb.append(mae)
        smape_list_hgb.append(smape)
        print(f'Parameter: {params}; MAE: {mae}; SMAPE: {smape}')

    print()

updrs_1: 2353 samples in training, 262 samples in testing.
Parameter: {'min_samples_leaf': 3, 'max_depth': 4}; MAE: 2.3746094566882503; SMAPE: 47.38371147954626
Parameter: {'min_samples_leaf': 10, 'max_depth': 2}; MAE: 2.849316418749054; SMAPE: 52.530380986646264
Parameter: {'min_samples_leaf': 20, 'max_depth': 16}; MAE: 2.377019893959594; SMAPE: 50.05473863101337
Parameter: {'min_samples_leaf': 5, 'max_depth': 16}; MAE: 2.3410959164268887; SMAPE: 48.31376332947447
Parameter: {'min_samples_leaf': 5, 'max_depth': 8}; MAE: 2.3130451999973185; SMAPE: 47.61535642424172
Parameter: {'min_samples_leaf': 10, 'max_depth': 32}; MAE: 2.3449763244762156; SMAPE: 49.17964347208073
Parameter: {'min_samples_leaf': 10, 'max_depth': 16}; MAE: 2.346066375155454; SMAPE: 49.153210708897824
Parameter: {'min_samples_leaf': 20, 'max_depth': 32}; MAE: 2.344057988015238; SMAPE: 49.832565058737025
Parameter: {'min_samples_leaf': 15, 'max_depth': 32}; MAE: 2.328084987291105; SMAPE: 48.0704497906605
Parameter: {'m

Like I've done previously, I will combine the parameters and their scores into a single dataframe for easy referencing.

In [83]:
HGBR_df = pd.concat([params_df_hgb,params_df_hgb,params_df_hgb,params_df_hgb], ignore_index=True)
HGBR_df['MAE'] = mae_list_hgb
HGBR_df['SMAPE'] = smape_list_hgb

HGBR_df

min_samples_leaf  max_depth       MAE       SMAPE
0                  3          4  2.374609   47.383711
1                 10          2  2.849316   52.530381
2                 20         16  2.377020   50.054739
3                  5         16  2.341096   48.313763
4                  5          8  2.313045   47.615356
5                 10         32  2.344976   49.179643
6                 10         16  2.346066   49.153211
7                 20         32  2.344058   49.832565
8                 15         32  2.328085   48.070450
9                 20          8  2.381437   49.868890
10                 3          4  2.537449   72.407854
11                10          2  3.200758   74.937118
12                20         16  2.216438   69.387499
13                 5         16  2.335312   69.995107
14                 5          8  2.346923   70.742559
15                10         32  2.294685   69.700333
16                10         16  2.309437   69.307048
17                20         32  2.252021   69.703308
18                15         32  2.332881   70.730807
19                20          8  2.254857   69.859487
20                 3          4  6.179714   58.143430
21                10          2  7.940158   64.494078
22                20         16  6.228545   61.868819
23                 5         16  6.126598   59.639585
24                 5          8  6.163945   59.579314
25                10         32  6.101021   60.989161
26                10         16  6.139752   61.037629
27                20         32  6.254724   61.592505
28                15         32  6.131740   60.328577
29                20          8  6.131106   59.916971
30                 3          4  1.004093  143.671064
31                10          2  1.207230  144.867232
32                20         16  0.877259  141.020589
33                 5         16  0.896476  139.778811
34                 5          8  0.889271  140.659416
35                10         32  0.893026  140.850302
36                10         16  0.896685  140.132824
37                20         32  0.882999  140.433113
38                15         32  0.885606  139.350251
39                20          8  0.910728  141.330534

Now onto AdaBoost Regressor with Decision Tree base regressor.

In [85]:
#ADABOOST REGRESSOR with decision tree regressor
X = big_data_imputed.drop(columns=targets, axis=1)

smape_list_ada = []
mae_list_ada = []

for target in targets:

    y = big_data_imputed[target]
    
    #splitting training and testing data. 
    X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.9, random_state=0)

    print(f'{target}: {len(X_train)} samples in training, {len(X_test)} samples in testing.')
    
    for params in random_params:
        #fitting and testing
        ada = AdaBoostRegressor(DecisionTreeRegressor(max_features=params['max_features'], max_depth=params['max_depth']), n_estimators=params['n_estimators'])
        ada.fit(X_train,y_train)    
        y_predict = ada.predict(X_test)

        #scoring
        mae = mean_absolute_error(y_test, y_predict)
        smape = smape_score(y_test, y_predict)
        print(f'Parameter: {params}; MAE: {mae}; SMAPE: {smape}')

        #saving scores in their lists
        mae_list_ada.append(mae)
        smape_list_ada.append(smape)

    print()

updrs_1: 2353 samples in training, 262 samples in testing.
Parameter: {'n_estimators': 500, 'max_features': 10, 'max_depth': 2}; MAE: 4.350973522315304; SMAPE: 66.8509207663236
Parameter: {'n_estimators': 200, 'max_features': 200, 'max_depth': 2}; MAE: 4.537534432111922; SMAPE: 68.20971317244128
Parameter: {'n_estimators': 500, 'max_features': 500, 'max_depth': 8}; MAE: 2.4267842131169495; SMAPE: 47.83117538636101
Parameter: {'n_estimators': 500, 'max_features': 25, 'max_depth': 8}; MAE: 2.7944045580838797; SMAPE: 51.89978339043074
Parameter: {'n_estimators': 100, 'max_features': 200, 'max_depth': 16}; MAE: 2.3487978307693576; SMAPE: 48.01331222585659
Parameter: {'n_estimators': 100, 'max_features': 200, 'max_depth': 4}; MAE: 3.313455984394416; SMAPE: 57.939083578091264
Parameter: {'n_estimators': 100, 'max_features': 100, 'max_depth': 8}; MAE: 2.6239332571559193; SMAPE: 49.91200052592894
Parameter: {'n_estimators': 200, 'max_features': 50, 'max_depth': 8}; MAE: 2.633784051302529; SMAP

In [86]:
ada_df = pd.concat([params_df,params_df,params_df,params_df], ignore_index=True)
ada_df['MAE'] = mae_list_ada
ada_df['SMAPE'] = smape_list_ada

ada_df

n_estimators  max_features  max_depth        MAE       SMAPE
0            500            10          2   4.350974   66.850921
1            200           200          2   4.537534   68.209713
2            500           500          8   2.426784   47.831175
3            500            25          8   2.794405   51.899783
4            100           200         16   2.348798   48.013312
5            100           200          4   3.313456   57.939084
6            100           100          8   2.623933   49.912001
7            200            50          8   2.633784   49.809392
8            100            50         32   2.564885   50.836950
9            500           500         32   2.511450   50.145761
10           500            10          2   5.262090   88.155134
11           200           200          2   5.549818   89.164451
12           500           500          8   2.632601   71.114798
13           500            25          8   3.182390   75.786834
14           100           200         16   2.142601   52.869207
15           100           200          4   4.145208   80.889232
16           100           100          8   2.790872   72.132984
17           200            50          8   2.951289   73.739028
18           100            50         32   2.217557   55.381728
19           500           500         32   2.167939   55.086096
20           500            10          2  12.596974   75.921348
21           200           200          2  12.018791   74.824739
22           500           500          8   6.576103   59.480627
23           500            25          8   8.452449   66.265418
24           100           200         16   5.983373   53.948106
25           100           200          4   9.794392   69.931940
26           100           100          8   7.214214   62.064628
27           200            50          8   7.781236   63.380957
28           100            50         32   6.193519   50.015804
29           500           500         32   5.890265   52.244061
30           500            10          2   2.916858  146.406700
31           200           200          2   2.546330  145.352208
32           500           500          8   1.273088  139.864766
33           500            25          8   1.567082  140.878939
34           100           200         16   0.829919   67.742876
35           100           200          4   1.912765  141.758666
36           100           100          8   1.285726  139.723102
37           200            50          8   1.388174  141.829294
38           100            50         32   0.805190   44.057150
39           500           500         32   0.790575   42.358871

In [87]:
#Tree based XGB Regressor using big_data_imputed

xgbr = xgboost.XGBRegressor()
smape_list_xgb = []
mae_list_xgb = []

X = big_data_imputed.drop(columns=targets, axis=1)

for target in targets:

    y = big_data_imputed[target]
    
    X_train, X_test, y_train, y_test = train_test_split(X.drop(columns='visit_id'), y, train_size=0.9, random_state=100) 

    print(f'{target}: {len(X_train)} samples in training, {len(X_test)} samples in testing.')
    
    for params in random_params:
        #fitting and testing
        xgbr.set_params(**params)
        xgbr.fit(X_train, y_train)
        y_predict = xgbr.predict(X_test)
        #scoring
        mae = mean_absolute_error(y_test, y_predict)
        smape = smape_score(y_test, y_predict)
        #saving scores in their lists
        mae_list_xgb.append(mae)
        smape_list_xgb.append(smape)

        print(f'Parameter: {params}; MAE: {mae}; SMAPE: {smape}')
        
    print()

updrs_1: 2353 samples in training, 262 samples in testing.
[15:05:36] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-0fdc6d574b9c0d168-1\xgboost\xgboost-ci-windows\src\learner.cc:767: 
Parameters: { "max_features" } are not used.

Parameter: {'n_estimators': 500, 'max_features': 10, 'max_depth': 2}; MAE: 2.356531773018473; SMAPE: 47.290864708579896
[15:05:56] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-0fdc6d574b9c0d168-1\xgboost\xgboost-ci-windows\src\learner.cc:767: 
Parameters: { "max_features" } are not used.

Parameter: {'n_estimators': 200, 'max_features': 200, 'max_depth': 2}; MAE: 2.3674468513265365; SMAPE: 47.174717218514814
[15:06:03] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-0fdc6d574b9c0d168-1\xgboost\xgboost-ci-windows\src\learner.cc:767: 
Parameters: { "max_features" } are not used.

Parameter: {'n_estimators': 500, 'max_features': 500, 'max_depth': 8}; MAE: 2.418668193586227

In [88]:
xgb_df = pd.concat([params_df,params_df,params_df,params_df], ignore_index=True)
xgb_df['MAE'] = mae_list_ada
xgb_df['SMAPE'] = smape_list_ada

xgb_df

n_estimators  max_features  max_depth        MAE       SMAPE
0            500            10          2   4.350974   66.850921
1            200           200          2   4.537534   68.209713
2            500           500          8   2.426784   47.831175
3            500            25          8   2.794405   51.899783
4            100           200         16   2.348798   48.013312
5            100           200          4   3.313456   57.939084
6            100           100          8   2.623933   49.912001
7            200            50          8   2.633784   49.809392
8            100            50         32   2.564885   50.836950
9            500           500         32   2.511450   50.145761
10           500            10          2   5.262090   88.155134
11           200           200          2   5.549818   89.164451
12           500           500          8   2.632601   71.114798
13           500            25          8   3.182390   75.786834
14           100           200         16   2.142601   52.869207
15           100           200          4   4.145208   80.889232
16           100           100          8   2.790872   72.132984
17           200            50          8   2.951289   73.739028
18           100            50         32   2.217557   55.381728
19           500           500         32   2.167939   55.086096
20           500            10          2  12.596974   75.921348
21           200           200          2  12.018791   74.824739
22           500           500          8   6.576103   59.480627
23           500            25          8   8.452449   66.265418
24           100           200         16   5.983373   53.948106
25           100           200          4   9.794392   69.931940
26           100           100          8   7.214214   62.064628
27           200            50          8   7.781236   63.380957
28           100            50         32   6.193519   50.015804
29           500           500         32   5.890265   52.244061
30           500            10          2   2.916858  146.406700
31           200           200          2   2.546330  145.352208
32           500           500          8   1.273088  139.864766
33           500            25          8   1.567082  140.878939
34           100           200         16   0.829919   67.742876
35           100           200          4   1.912765  141.758666
36           100           100          8   1.285726  139.723102
37           200            50          8   1.388174  141.829294
38           100            50         32   0.805190   44.057150
39           500           500         32   0.790575   42.358871